// mrf-utils/index.js
var mongo = require('mongodb');
var ObjectId = mongo.ObjectId;
var mongoSanitize = require('express-mongo-sanitize');

module.exports = function(app){
	helpers = {
		checkIn: function(json, fields, callback){
			fields.forEach(function(elem){
				if(!(elem in json)){
					callback(elem, false);
				} else{
					callback(elem, true);
				}
			});
		},

		isJSON: function(item){
			if(item == null) return false;
			return typeof item === "object" && item.constructor === {}.constructor;
		},

		isInt: function(item){
			return !(isNaN(item) || parseInt(Number(item)) != item || isNaN(parseInt(item, 10)));
		},

		toBool: function(bool){
			if(typeof bool === 'boolean') return bool;
			if(bool === "true") return true;
			if(bool === "false") return false;
		},

		isoFormat: "YYYY-MM-DDTHH:mm:ssZ",
		mrfDefaults: function(club_id, division_id, year, month){
			return {
				club_id: club_id,
				division_id: division_id,
				year: year,
				month: month,
				status: 0,
				submissionTime: null,
				updates: {
					duesPaid: null,
					newDuesPaid: null
				},
				goals: [],
				meetings:[
					{
						date: null,
						numMembers: null,
						numNonHomeMembers: null,
						numKiwanis: null,
						numGuests: null,
						advisorAttended:{
							faculty: null,
							kiwanis: null
						}
					},

					{
						date: null,
						numMembers: null,
						numNonHomeMembers: null,
						numKiwanis: null,
						numGuests: null,
						advisorAttended:{
							faculty: null,
							kiwanis: null
						}
					},

					{
						date: null,
						numMembers: null,
						numNonHomeMembers: null,
						numKiwanis: null,
						numGuests: null,
						advisorAttended:{
							faculty: null,
							kiwanis: null
						}
					},

					{
						date: null,
						numMembers: null,
						numNonHomeMembers: null,
						numKiwanis: null,
						numGuests: null,
						advisorAttended:{
							faculty: null,
							kiwanis: null
						}
					},

					{
						date: null,
						numMembers: null,
						numNonHomeMembers: null,
						numKiwanis: null,
						numGuests: null,
						advisorAttended:{
							faculty: null,
							kiwanis: null
						}
					},
				],

				dcm:{
					date: null,
					presidentAttended: null,
					numMembers: null,
					nextDate: null
				},

				communications:{
					ltg:{
						message: null,
						contacted:{
							visit: null,
							phone: null,
							email: null,
							newsletter: null,
							other: null
						}
					},
					dboard: null
				},

				kfamReport: null,
				labels: [],
				color: null
			}
		},

		dotNotate: function(obj,target,prefix) {
			target = target || {},
			prefix = prefix || "";

			Object.keys(obj).forEach(function(key) {
				if(helpers.isJSON(obj[key])) {
					helpers.dotNotate(obj[key], target, prefix + key + ".");
				}else if(Array.isArray(obj[key])){
					for(var i in obj[key]){
						if(helpers.isJSON(obj[key][i])){
							helpers.dotNotate(obj[key][i], target, prefix + key + "." + i + ".");
						}else{
							target[prefix + key + "." + i] = obj[key][i];
						}
					}
				}else{
					target[prefix + key] = obj[key];
				}
			});

			return target;
		}
	}

	return helpers;
}
